framework/src/loops/ComputeMortarFunctor.C

```diff
@@ -82,175 +80,227 @@ ComputeMortarFunctor::operator()()
+          // Map quadrature points to secondary side
+          const Real xi1_eta = 0.5 * (1 - eta) * msinfo.xi1_a + 0.5 * (1 + eta) * msinfo.xi1_b;
+          secondary_xi_pts.push_back(xi1_eta);
+
+          // Map quadrature points to primary side
+          const Real xi2_eta = 0.5 * (1 - eta) * msinfo.xi2_a + 0.5 * (1 + eta) * msinfo.xi2_b;
+          primary_xi_pts.push_back(xi2_eta);
+        }
+      }
+      else
       {
-        Real eta = _qrule_msm->qp(qp)(0);
-        Real xi1_eta = 0.5 * (1 - eta) * msinfo.xi1_a + 0.5 * (1 + eta) * msinfo.xi1_b;
-        custom_xi1_pts[qp] = xi1_eta;
+        // Now that has_primary logic gone, could combine these to make more efficient
+        projectQPoints3d(msm_elem, msinfo.secondary_elem, secondary_sub_elem_ind, secondary_xi_pts);
+        projectQPoints3d(msm_elem, msinfo.primary_elem, primary_sub_elem_ind, primary_xi_pts);
       }
-    }
-    else
-      projectQPoints3d(msm_elem, msinfo.secondary_elem, secondary_sub_elem_ind, custom_xi1_pts);
-
-    const auto normals = _amg.getNormals(*msinfo.secondary_elem, custom_xi1_pts);
-    const auto nodal_normals = _amg.getNodalNormals(*msinfo.secondary_elem);
 
-    const Elem * reinit_secondary_elem = secondary_ip;
+      // If edge dropping case we need JxW on the msm to compute dual shape functions
+      if (_assembly.needDual() /*&& edge_dropping*/)
+        std::copy(std::begin(_JxW_msm), std::end(_JxW_msm), std::back_inserter(JxW));
+
+#ifndef NDEBUG
+      // Verify that the expected number of quadrature points have been inserted
+      expected_length += _qrule_msm->n_points();
+      mooseAssert(secondary_xi_pts.size() == expected_length,
+                  "Fewer than expected secondary quadrature points");
+      mooseAssert(primary_xi_pts.size() == expected_length,
+                  "Fewer than expected primary quadrature points");
+
+      if (_assembly.needDual())
+        mooseAssert(JxW.size() == expected_length, "Fewer than expected JxW values computed");
+#endif
+    }
 
-    // If we're on the displaced mesh, we need to get the corresponding undisplaced elem before
-    // calling _fe_problem.reinitElemFaceRef
-    if (_displaced)
-      reinit_secondary_elem = _fe_problem.mesh().elemPtr(reinit_secondary_elem->id());
+    // Note that if we aren't on an edge-dropping element it would be better to assemble
+    // with standard quadrature. Detecting edge dropping is a bit tricky though, in 3D we
+    // can't simply compare volume of mortar segments to secondary since the linearized elements
+    // don't have the same volume as the warped secondary elements. In 3D we could get the info
+    // from mortar_segment_helper->remainder (would need to add up for all sub-elements). In 2D
+    // we could just compare volumes like mentioned before. This only matters if we decide to
+    // treat non-edge dropping (interior) elements different than edge dropping ones.
 
```

Comment:
@jbadger95 Pretty sure you have explained this before, but, here it goes... What would be the advantage(s) of assembling internal elements with standard quadratures?

Reply:
By the way, great comment.